Add barrier() at the end of init_process_group and new_group.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45181 Add barrier() at the end of init_process_group and new_group.**

`init_process_group` and `new_group` update a bunch of global
variables after initializing the actual process group. As a result, there is a
race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

To resolve this issue, I've added barrier() at the end of both of these calls.
This ensures that once these calls return we are guaranteed about correct
initialization on all ranks.

Since these calls are usually done mostly during initialization, it should be
fine to add the overhead of a barrier() here.

#Closes: https://github.com/pytorch/pytorch/issues/40434, https://github.com/pytorch/pytorch/issues/40378

Differential Revision: [D23858025](https://our.internmc.facebook.com/intern/diff/D23858025/)

pritamdamania committed:
Add barrier() at the end of init_process_group and new_group.

`init_process_group` and `new_group` update a bunch of global
variables after initializing the actual process group. As a result, there is a
race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

To resolve this issue, I've added barrier() at the end of both of these calls.
This ensures that once these calls return we are guaranteed about correct
initialization on all ranks.

Since these calls are usually done mostly during initialization, it should be
fine to add the overhead of a barrier() here.

#Closes: https://github.com/pytorch/pytorch/issues/40434, https://github.com/pytorch/pytorch/issues/40378

Differential Revision: [D23858025](https://our.internmc.facebook.com/intern/diff/D23858025/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -436,6 +436,10 @@ def init_process_group(backend,
+    # barrier at the end to ensure that once we return from this method, all
+    # process groups including global variables are updated correctly on all
+    # ranks.
+    barrier()
```
We might need to guard this with try-except or if-else as it's possible that 3rd party extensions for c10d does not support barrier. See #45126

pritamdamania87 replied:
I'm wondering how 3rd party extensions would work without supporting barrier? Currently ProcessGroup.hpp clearly defines the `barrier` method as pure virtual and as a result we should probably require extensions to support this. Do you know if there are any existing 3rd party extensions that do not support this? It looks like oneCCL and torch-ucc do support barrier().

mrshenli replied:
IIRC, we have one test c10d extension whose `barrier()` implementation just throws an unsupported error.

mrshenli reviewed (approved):
Stamp to unblock.

lw commented:
How did you check that this fixes those failures? The OSX job on CircleCI is failing (for another reason) and from its logs it seems it's not running the DdpUnderDistAutogradTest suite.

pritamdamania87 commented:
> How did you check that this fixes those failures? The OSX job on CircleCI is failing (for another reason) and from its logs it seems it's not running the DdpUnderDistAutogradTest suite.

Shen pointed me to this: https://circleci.com/docs/2.0/ssh-access-jobs/. Basically you re-run an existing CircleCI job with ssh enabled. So I sshed into the vm, and ran the test 100 times and some runs failed. Then I locally applied the fix in this PR and ran the test 100 times again and all runs passed.

pritamdamania committed:
Update on "Add barrier() at the end of init_process_group and new_group."


`init_process_group` and `new_group` update a bunch of global
variables after initializing the actual process group. As a result, there is a
race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

To resolve this issue, I've added barrier() at the end of both of these calls.
This ensures that once these calls return we are guaranteed about correct
initialization on all ranks.

Since these calls are usually done mostly during initialization, it should be
fine to add the overhead of a barrier() here.

#Closes: https://github.com/pytorch/pytorch/issues/40434, https://github.com/pytorch/pytorch/issues/40378

Differential Revision: [D23858025](https://our.internmc.facebook.com/intern/diff/D23858025/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/distributed_c10d.py:
```diff
@@ -436,6 +436,10 @@ def init_process_group(backend,
+    # barrier at the end to ensure that once we return from this method, all
+    # process groups including global variables are updated correctly on all
+    # ranks.
+    barrier()
```
cc @agolynski FYI, after this change, c10d extensions cannot throw in `barrier()`, otherwise they won't be able to initialize default process groups.

pritamdamania committed:
Update on "Add barrier() at the end of init_process_group and new_group."


`init_process_group` and `new_group` update a bunch of global
variables after initializing the actual process group. As a result, there is a
race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

To resolve this issue, I've added barrier() at the end of both of these calls.
This ensures that once these calls return we are guaranteed about correct
initialization on all ranks.

Since these calls are usually done mostly during initialization, it should be
fine to add the overhead of a barrier() here.

#Closes: https://github.com/pytorch/pytorch/issues/40434, https://github.com/pytorch/pytorch/issues/40378

Differential Revision: [D23858025](https://our.internmc.facebook.com/intern/diff/D23858025/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add barrier() at the end of init_process_group and new_group."


`init_process_group` and `new_group` update a bunch of global
variables after initializing the actual process group. As a result, there is a
race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

To resolve this issue, I've added barrier() at the end of both of these calls.
This ensures that once these calls return we are guaranteed about correct
initialization on all ranks.

Since these calls are usually done mostly during initialization, it should be
fine to add the overhead of a barrier() here.

#Closes: https://github.com/pytorch/pytorch/issues/40434, https://github.com/pytorch/pytorch/issues/40378

Differential Revision: [D23858025](https://our.internmc.facebook.com/intern/diff/D23858025/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add barrier() at the end of init_process_group and new_group."


`init_process_group` and `new_group` update a bunch of global
variables after initializing the actual process group. As a result, there is a
race that after initializing the process group on say rank 0, if we immediately
check the default process group on rank 1 (say via RPC), we might actually get
an error since rank 1 hasn't yet updated its _default_pg variable.

To resolve this issue, I've added barrier() at the end of both of these calls.
This ensures that once these calls return we are guaranteed about correct
initialization on all ranks.

Since these calls are usually done mostly during initialization, it should be
fine to add the overhead of a barrier() here.

#Closes: https://github.com/pytorch/pytorch/issues/40434, https://github.com/pytorch/pytorch/issues/40378

Differential Revision: [D23858025](https://our.internmc.facebook.com/intern/diff/D23858025/)

[ghstack-poisoned]
